hibernate-core/src/main/java/org/hibernate/loader/Loader.java

```diff
@@ -68,6 +68,7 @@
 import org.hibernate.event.spi.PostLoadEvent;
```

Comment:
It seems a good catch. I am the community developer who implemented 'fetch graph' in v5. Also I implemented entity graph fully in v6. Yeah, I agree. The current implementation is buggy and difficult to maintain. That is because v5 infrastructure and design makes it difficult to implement and that is why 'fetch graph' was not implemented for a long time. The status has been changed in v6 and I think we have implemented the whole entity graph elegantly there already.

I would take a look at the PR and thanks for the two options to choose from. However, I am community contributor and have very limited free cycles out of my daily job. Seems an endless fixing of entity graph in v5, :(.

Reply:
Right, I also noticed that a better solution to handle fetch graphs would require very significant changes to ORM 5. That's why I didn't even try :)

Reply:
I would prefer non-intrusive approach to fix the issue without much performance cost. The improvement in v6 is we have complete access to the navigation role path of the entity to be loaded. There might be such role info already in TwoPhaseLoad, though no time to investigate.

Reply:
Would you mind my creating another PR if I can solve the issue without changing Loader or creating any new data structure (reusing existing info instead)? Not necessarily possible. Just in case.

Reply:
I wouldn't mind as long as my tests still pass, but I'd rather you create another ticket and a follow-up PR if you want to make more changes to address more problems. You know, one step at a time.

I also think it might not be the best use of your time, since 6.0 is the future, but that's _your_ time :)

Reply:
I tried to create my PR but failed. Especially there is no way to even know some hydrated object is of root entity or not (the relationship among them is totally lost in TwoPhaseLoad class). Seems no better way to solve the issue. Great job and thanks a ton.